src/config/prettier-config-explorer/config-searcher.js

```diff
@@ -0,0 +1,91 @@
+import path from "node:path";
+import iterateDirectoryUp from "iterate-directory-up";
+import {
+  CONFIG_FILE_NAMES,
+  createCachedFunction,
+  fileExists,
+} from "./common.js";
+import { loadConfigFromPackageJson } from "./loaders.js";
+
+async function isPackageJsonFileWithPrettierConfig(file) {
+  try {
+    return Boolean(await loadConfigFromPackageJson(file));
+  } catch {
+    return false;
+  }
+}
+
+async function searchConfigInDirectory(directory) {
+  for (const fileName of CONFIG_FILE_NAMES) {
+    const file = path.join(directory, fileName);
+
+    if (!(await fileExists(file))) {
+      continue;
+    }
+
+    if (
+      fileName !== "package.json" ||
+      (await isPackageJsonFileWithPrettierConfig(file))
+    ) {
+      return file;
+    }
+  }
+}
+
+/** @type {Map} */ // @ts-expect-error -- intentionally not add the `get` method
+const noopMap = { has: () => false, set() {} };
+class Searcher {
+  #cache;
```

Comment:
Is this a just boolean value? If true I prefer naming as `shouldCache`. To clarify whether this is an object that has cached values or a just boolean:

```suggestion
  #shouldCache;
```